p2p/net/swarm/swarm.go

```diff
@@ -646,10 +650,34 @@ func isDirectConn(c *Conn) bool {
 // To check if we have an open connection, use `s.Connectedness(p) ==
 // network.Connected`.
 func (s *Swarm) Connectedness(p peer.ID) network.Connectedness {
-	if s.bestConnToPeer(p) != nil {
-		return network.Connected
+	s.conns.RLock()
+	defer s.conns.RUnlock()
+
+	return s.connectednessUnlocked(p, false)
+}
+
+// connectednessUnlocked returns the connectedness of a peer. For sending peer connectedness
+// changed notifications consider closed connections. When remote closes a connection
+// the underlying transport connection is closed first, so tracking changes to the connectedness
+// state requires considering this recently closed connections impact on Connectedness.
+func (s *Swarm) connectednessUnlocked(p peer.ID, considerClosed bool) network.Connectedness {
```

Comment:
We may also have limited connections to the peer. So once a connection is closed, we check:
* If there is at least 1 unlimited connection, return `Connected`
* Else, if there is at least 1 limited connection, return `Limited`
* Else, return `NotConnected`

> Why are we potentially returning `Connected` in the case that every conn in the map returns `c.IsClosed()=true` with `considerClosed=true`

It is true. However, `considerClosed` is `true` only when called from `addConn` and `removeConn`, which hold the lock. `removeConn` makes sure to remove the connection from `s.conns.m`, so the function will not return `Connected` if we removed the last unlimited connection.

Though, I don't really understand why `considerClosed` is needed at all. @sukunrt could you provide some context?

Reply:
I agree this is confusing. 

There are two different states this function is called from: 

When closing the connection from our side by calling Close, this function is called when the underlying transport connection is not closed. 

When closing the connection because the remote Closed the connection, this function is called when the underlying transport connection has already been closed.

Reply:
Consider that we are connected to the peer over a QUIC and a relayed connection. 

Now the peer closes the QUIC connection. So we have moved from a `Connected` state to `Limited` connectivity state. 

To make this work, when calculating oldState in removeConn we need to consider connectivity assuming QUIC connection was still connected.

Alternatively we can store the last sent connectedness event for a peer. That would make this calculation cleaner at the expense of more state for an edge case.

Reply:
> Alternatively we can store the last sent connectedness event for a peer. That would make this calculation cleaner at the expense of more state for an edge case.

I think we should do this.